lib/active_merchant/billing/gateways/adyen.rb

```diff
@@ -246,6 +248,45 @@ def add_extra_data(post, payment, options)
         add_merchant_data(post, options)
       end
 
+      def extract_and_transform(mapper, from)
+        mapper.each_with_object({}) do |key_map, hsh|
+          key, item_key = key_map[0], key_map[1]
+          hsh[key] = from[item_key.to_sym]
+        end
+      end
+
+      def add_level_2_data(post, options)
+        mapper = {
```

Comment:
I just had a thought - given the complexity of these methods, it would be good to begin them with a guard clause that will exit early if there is no level 2/3 data available.
```suggestion
        return unless options[:level_2_data]

        mapper = {
```

Reply:
I guess for this one it is slightly more complex due to the tax and cutomer reference fields not being contained in the level_2_data subhash...

Reply:
Ok, perfect. I understood the point, you are right it would be better to have it. 

note:
early return was added.